Sources/Sentry/SentryFileManager.m

```diff
@@ -634,23 +634,132 @@ - (BOOL)isDirectory:(NSString *)path
+    NSString *lastPathComponent = [[[NSBundle mainBundle] executablePath] lastPathComponent];
+
+    // Due to `NSProcessInfo` and `NSBundle` not being mockable in unit tests, we extract only the
+    // logic to a separate function.
+    return sentryBuildScopedCachesDirectoryPath(
+        cachesDirectory, isSandboxed, bundleIdentifier, lastPathComponent);
+#endif
+}
+
+NSString *_Nullable sentryBuildScopedCachesDirectoryPath(NSString *cachesDirectory,
+    BOOL isSandboxed, NSString *_Nullable bundleIdentifier, NSString *_Nullable lastPathComponent)
+{
+    // If the app is sandboxed, we can just use the provided caches directory.
+    if (isSandboxed) {
+        return cachesDirectory;
+    }
+
+    // If the macOS app is not sandboxed, we need to manually create a scoped cache
+    // directory. The cache path must be unique an stable over app launches, therefore we
+    // can not use any changing identifier.
+    SENTRY_LOG_DEBUG(
+        @"App is not sandboxed, extending default cache directory with bundle identifier.");
+
+    if (lastPathComponent.length > 0) {
+        // If we've already used a scoped cache directory path with the executable name in the past
+        // due to a nil or zero-length bundle identifier, we'll stick with that regardless of what
+        // is in the app identifier now.
+        NSString *executableNameBasedPath =
+            [cachesDirectory stringByAppendingPathComponent:lastPathComponent];
+        BOOL isDirectory = NO;
+        if ([NSFileManager.defaultManager fileExistsAtPath:executableNameBasedPath
+                                               isDirectory:&isDirectory]
+            && isDirectory) {
+            return executableNameBasedPath;
+        }
+    }
+
+    NSString *_Nullable identifier = bundleIdentifier;
+    if (identifier == nil) {
+        SENTRY_LOG_WARN(@"No bundle identifier found, using main bundle executable name.");
+        identifier = lastPathComponent;
+    } else if (identifier.length == 0) {
+        SENTRY_LOG_WARN(@"Bundle identifier exists but is zero length, using main bundle "
+                        @"executable name.");
+        identifier = lastPathComponent;
+    }
+
+    // If neither the bundle identifier nor the executable name are available, we can't
+    // create a unique and stable cache directory.
+    // We do not fall back to any default path, because it could be shared with other apps
+    // and cause leaks impacting other apps.
+    if (identifier == nil) {
+        SENTRY_LOG_ERROR(@"No bundle identifier found, cannot create cache directory.");
+        return nil;
+    }
+
+    // It's unlikely that the executable name will be zero length, but we'll cover this case anyways
+    if (identifier.length == 0) {
+        SENTRY_LOG_ERROR(@"Executable name was zero length.");
+        return nil;
+    }
+
+    return [cachesDirectory stringByAppendingPathComponent:identifier];
+}
```

Comment:
`l`: This code has little to do with the rest of the FileManager. We could put it into an extra file to not bloat the FileManager even more.

Reply:
Yes, but in another PR please.